### ISSUE TITLE
Reusing EVP_CIPHER_CTX on ZLinux does not return same ciphertext for same input

### ISSUE THREAD

**Issue by ashwanikadian** [openssl/openssl#23175]
We are trying to reuse an EVP_CIPHER_CTX object for multiple doing encryption/decryption operations in multiple iterations. The Key and IV doe snot change for us in each of those iterations.

After creating and initializing an EVP_CIPHER_CTX object using "EVP_CIPHER_CTX_new()", followed by  "EVP_EncryptInit(cipher_ctx, cipher, key_data, iv_data)" call, we call "EVP_CipherInit(cipher_ctx, NULL, NULL, NULL, -1)" to reset the EVP_CIPHER_OBJECT for the next encryption/decryption operation and then call EVP_EncryptUpdate() and EVP_EncryptFinal() APIs to do the encryption in each iteration. 

I observed that it works perfectly fine on Linux and other platforms for all AES modes. But specifically on ZLinux with AES-CFB mode, we have an issue that, we are getting different cipher text in each iteration, for the same input text. The key and IV remains the same in each iterations. Linux and other platforms are returning same cipher text for AES_CFB mode as expected. On ZLinux also, XTS, GCM and other modes are also returning the same cipher text as expected. So, this is the problem specific to ZLinux and specific to AES-CFB mode.

Another observation was that, if I pass same "key_data" in my reset call as mentioned above, (like, EVP_CipherInit(cipher_ctx, NULL, key_data, NULL, -1)), instead of passing NULL for the key data, then I start getting the same cipher text as expected.

I checked the assembler code for AES-CFB for ZLinux in file "providers/implementations/ciphers/cipher_aes_hw_s390x.inc", and found that we are setting "adat->plat.s390x.res" to "0" in s390x_aes_cfb128_initkey() API, which gets called by the reset call EVP_CipherInit(). But after doing the encryption operation in API s390x_aes_cfb128_cipher_hw(), we are setting "adat->plat.s390x.res" to "n". I tested and confirmed that this is causing the problem I explained above. If I set "adat->plat.s390x.res" to "0" in s390x_aes_cfb128_cipher_hw(), then I start getting the expected results on ZLinux with AES-CFB.

Can you please review the change below in file "providers/implementations/ciphers/cipher_aes_hw_s390x.inc" which helped me to get the correct expected results and merge it to the OpenSSL codeline if it looks good or please suggest a fix for this issue and the reason for keeping "adat->plat.s390x.res" value as "n" in s390x_aes_cfb128_cipher_hw(). Please let me know if you think that the below change may break any other usecase.

``` diff
@@ -164,7 +164,7 @@
     }
 
     memcpy(dat->iv, adat->plat.s390x.param.kmo_kmf.cv, dat->ivlen);
-    adat->plat.s390x.res = 0;
+    adat->plat.s390x.res = n;
     return 1;
 }

```

**Comment by jamuir** [comment]
Thanks for reporting this issue.

It is difficult to read your patch from your comment above since you did not surround it with triple-backticks.

I think this is what you suggested:

```diff
diff --git a/providers/implementations/ciphers/cipher_aes_hw_s390x.inc b/providers/implementations/ciphers/cipher_aes_hw_s390x.inc
index c8282dbd08..1d82e3e1a1 100644
--- a/providers/implementations/ciphers/cipher_aes_hw_s390x.inc
+++ b/providers/implementations/ciphers/cipher_aes_hw_s390x.inc
@@ -164,7 +164,7 @@ static int s390x_aes_cfb128_cipher_hw(PROV_CIPHER_CTX *dat, unsigned char *out,
     }
 
     memcpy(dat->iv, adat->plat.s390x.param.kmo_kmf.cv, dat->ivlen);
-    adat->plat.s390x.res = n;
+    adat->plat.s390x.res = 0;
     return 1;
 }

```

We will review your change and report back here.

**Comment by levitte** [comment]
I edited the original post to add markers for the diff section.  Hope that clarifies things

**Comment by t8m** [comment]
@holger-dengler @ifranzki Could you please look at this?

**Comment by holger-dengler** [comment]
ok, I'll try to understand whats going wrong.

**Comment by holger-dengler** [comment]
After a first look, I'm not sure, that the proposed fix is the solution for this problem. We had a similar issue in 2021 with the iv in the hw-specific part of the context (commit 732a4d15b0da7c04437ea828b2915a691b6e38db).

The proposed fix may work for single-shot messages, but it will cause problems for multi-chunk cases. I'll try write some code, reproduce the problem and try to develop a proper fix for it.

**Comment by holger-dengler** [comment]
@ashwanikadian According to the man Page, `EVP_CIPHER_CTX_reset()` has to be called on re-using the context. 

From `man EVP_ENCRYPTINIT`:
```
...
EVP_CIPHER_CTX_reset()
    Clears all information from a cipher context and free up any allocated memory associated with it,  except  the  ctx  itself.  This  function
    should be called anytime ctx is reused by another EVP_CipherInit() / EVP_CipherUpdate() / EVP_CipherFinal() series of calls.
...
```

Are you aware of that?

**Comment by t8m** [comment]
@holger-dengler IMO that documentation is fairly outdated. We support calling EVP_CIPHER_CTX_init() without _reset().

However in practical use-cases you always reinitialize at least with a different IV and not call EVP_CIPHER_CTX_init()  with all NULL arguments. Nevertheless it would be better to support even this kind of reinitialization if other implementations support it.

**Comment by holger-dengler** [comment]
@t8m Thanks for this hint. It looks like we're tracking partial processed blocks in the platform-specific part of the context. I've currently no idea, how this information can be reset with the current set of registered provider functions.

**Comment by holger-dengler** [comment]
@ashwanikadian Can you please have a look at the following test code? Is this similar to your usecase?

If I run this test code below on my test env (s390x, openssl 3.0.9), I get the same ciphertext in all iterations. I also get the same results on my local machine (x86_64, openssl 3.1.4). Which openssl version do you use? There was a bug in the s390x-specific code, which has been fixed by commit 732a4d15b0da7c04437ea828b2915a691b6e38db. Can you check, if your openssl version has this commit? If not, can you please check, if the commit is fixing your problem?

cfb.c
```
#include <stdio.h>
#include <string.h>
#include <openssl/evp.h>

#define AES_BSIZE	16

static const unsigned char
key[16] = { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
	    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };

static const unsigned char
iv[16] =  { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
	    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };

static const unsigned char
plain[] = "This is the plain text";

static unsigned char
cipher[sizeof(plain)];

void hexdump(const char *msg, const unsigned char *b, size_t blen)
{
	size_t i, part;

	if (msg)
		printf("%s\n", msg);

	for (i = 0, part = 0; i < blen; i++) {
		printf("0x%02x ", (int)(b[i] & 0xff));
		part = (part + 1) % 8;
		if (!part)
			printf("\n");
	}
	if (part)
		printf("\n");
}

int main(void)
{
	EVP_CIPHER_CTX *ctx;
	int outl, i;

	memset(cipher, 0, sizeof(cipher));

	printf("init:\n");
	hexdump("plain", plain, sizeof(plain));
	hexdump("cipher", cipher, sizeof(cipher));

	ctx = EVP_CIPHER_CTX_new();
	if (!ctx) {
		fprintf(stderr, "EVP_CIPHER_CTX_new failed\n");
		return(99);
	}
	if (!EVP_EncryptInit(ctx, EVP_aes_128_cfb8(), key, iv)) {
		fprintf(stderr, "EVP_EncryptInit failed\n");
		return(99);
	}

	for (i = 1; i < 5; i++) {
		memset(cipher, 0, sizeof(cipher));
		if (!EVP_CipherInit(ctx, NULL, NULL, NULL, -1)) {
			fprintf(stderr, "EVP_CipherInit failed\n");
			return(i);
		}
		if (!EVP_EncryptUpdate(ctx, cipher, &outl, plain, sizeof(plain))) {
			fprintf(stderr, "EVP_EncryptUpdate failed\n");
			return(i);
		}
		if (!EVP_EncryptFinal(ctx, cipher, &outl)) {
			fprintf(stderr, "EVP_EncryptFinal failed\n");
			return(2);
		}
		printf("\nround %d:\n", i);
		hexdump("cipher", cipher, sizeof(cipher));
	}
	return 0;
}
```

Output:
```
$ uname -a
Linux a3545040.lnxne.boe 6.5.5-100.fc37.s390x #1 SMP Sat Sep 23 22:29:55 UTC 2023 s390x s390x s390x GNU/Linux
$ ./cfb
init:
plain
0x54 0x68 0x69 0x73 0x20 0x69 0x73 0x20
0x74 0x68 0x65 0x20 0x70 0x6c 0x61 0x69
0x6e 0x20 0x74 0x65 0x78 0x74 0x00
cipher
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
0x00 0x00 0x00 0x00 0x00 0x00 0x00

round 1:
cipher
0x5e 0xb4 0x4a 0x15 0x57 0x7e 0xd9 0xf3
0x80 0x33 0x2d 0x7f 0xd1 0xcf 0xd1 0xb2
0x0e 0xfc 0x9e 0x40 0x85 0x1e 0x6d

round 2:
cipher
0x5e 0xb4 0x4a 0x15 0x57 0x7e 0xd9 0xf3
0x80 0x33 0x2d 0x7f 0xd1 0xcf 0xd1 0xb2
0x0e 0xfc 0x9e 0x40 0x85 0x1e 0x6d

round 3:
cipher
0x5e 0xb4 0x4a 0x15 0x57 0x7e 0xd9 0xf3
0x80 0x33 0x2d 0x7f 0xd1 0xcf 0xd1 0xb2
0x0e 0xfc 0x9e 0x40 0x85 0x1e 0x6d

round 4:
cipher
0x5e 0xb4 0x4a 0x15 0x57 0x7e 0xd9 0xf3
0x80 0x33 0x2d 0x7f 0xd1 0xcf 0xd1 0xb2
0x0e 0xfc 0x9e 0x40 0x85 0x1e 0x6d
```

**Comment by ashwanikadian** [comment]
Hi @t8m and @holger-dengler 

Thanks for you comments.

@holger-dengler I am using OpenSSL 3.0.12 for my testing. I verified that the commit you mentioned is already available in OpenSSL 3.0.12 code.

I tested you code and found it to be returning the same ciphertext as you also mentioned. But then I compared that with our code found that the difference is only in the EVP_CIPHER object.

You are fetching the cipher object EVP_aes_128_cfb8(), but in our code, we are fetching it like, EVP_CIPHER_fetch(NULL, "AES-128-CFB", NULL) and EVP_CIPHER_fetch(NULL, EVP_CIPHER_name(EVP_aes_128_cfb()), NULL). The ciphers fetched like the one mentioned now are not working as I had mentioned before.

So, if you change your following line 

if (!EVP_EncryptInit(ctx, EVP_aes_128_cfb8(), key, iv)) {

to 

if (!EVP_EncryptInit(ctx, EVP_CIPHER_fetch(NULL, "AES-128-CFB", NULL), key, iv)) {

then it starts failing with OpenSSL 3.0.12.
